syntax = "proto3";

import "google/protobuf/empty.proto";


package service;

// Definição do tipo complexo
message ComplexType {
  string name = 1;
  int32 age = 2;
}

// Definições das mensagens de requisição e resposta
message LongRequest {
  int64 value = 1;
}

message LongResponse {
  int64 value = 1;
}

message EightLongsRequest {
  int64 v1 = 1;
  int64 v2 = 2;
  int64 v3 = 3;
  int64 v4 = 4;
  int64 v5 = 5;
  int64 v6 = 6;
  int64 v7 = 7;
  int64 v8 = 8;
}

message StringRequest {
  string value = 1;
}

message StringResponse {
  string value = 1;
}

message ComplexRequest {
  ComplexType complex = 1;
}

message ComplexResponse {
  ComplexType complex = 1;
}

// Serviço
service TestService {
  rpc NoArgsNoReturn (google.protobuf.Empty) returns (google.protobuf.Empty);
  rpc OneLong (LongRequest) returns (LongResponse);
  rpc EightLongs (EightLongsRequest) returns (LongResponse);
  rpc OneString (StringRequest) returns (StringResponse);
  rpc ComplexOperation (ComplexRequest) returns (ComplexResponse);
}
